Show pull requests against branches

jesseduffield opened the pull request:
- **PR Description**

If the user has `gh` installed and is logged in (`gh auth login`), lazygit shows GitHub PR icons next to the names of branches that have an associated PR, colored by the PR's status (green=open, red=closed, purple=merged).

Selecting a branch and pressing `shift-G` opens the PR in the browser.

stefanhaller commented:
Oooh, very nice!

A few thoughts after trying it (very briefly though, so I might be missing a lot of things):
- It looks like it's fetching the PRs only after it's done fetching the remote(s). Is there a reason for this? For me, fetching often takes a lot of time, so I have to wait quite a while until the PR numbers show up.
- I'm not sure the 500 limit works well. For our monorepo at work this amounts to roughly one month; sometimes we have PRs open for longer than that, so some of my open branches don't show PR numbers. Also, it would be really nice if you would also see PR numbers for very old branches. I haven't thought at all about how to achieve that, so sorry if this is not very constructive input.
- It looks like PR numbers are only shown in the local branches panel. I'm thinking they could be even more useful in the remote branches list. (Not totally sure about that, it's just a feeling for now.)
- Do I understand it right that at the moment the only thing this does is add the PR numbers to the display? I somehow expected you would be able to do more with this; the obvious wish is to open a PR in the web browser (I have a custom command for this). Maybe the `o` command could be overloaded for this? Show the existing PR if we know there is one, and open a new one if not.

jesseduffield commented:
> It looks like it's fetching the PRs only after it's done fetching the remote(s). Is there a reason for this? For me, fetching often takes a lot of time, so I have to wait quite a while until the PR numbers show up.

It needs remotes to map from PRs back to branches i.e. the PR has an owner e.g. 'jesseduffield') and the remote has a URL (e.g. `github.com/jesseduffield/lazygit`) and the branch has the name of the remote. So we need to piece those together to know that a particular PR should be assigned to a particular branch. imo getting remotes should be really fast because it's just a matter of reading a file so if it's currently a bottleneck it's worth looking into.

> I'm not sure the 500 limit works well. For our monorepo at work this amounts to roughly one month; sometimes we have PRs open for longer than that, so some of my open branches don't show PR numbers. Also, it would be really nice if you would also see PR numbers for very old branches. I haven't thought at all about how to achieve that, so sorry if this is not very constructive input.

No need to apologise, I've been thinking the same thing. I wish there was a way to pass multiple branch heads to gh or to the graphql endpoint (gh is just a way of talking to graphql) but from what I've seen and tried, there's not. We could try a parallelised approach of shelling out to a bunch of gh processes at once but it would probably be rate limited.

> It looks like PR numbers are only shown in the local branches panel. I'm thinking they could be even more useful in the remote branches list. (Not totally sure about that, it's just a feeling for now.)

I agree, and it shouldn't be hard to support that.

> Do I understand it right that at the moment the only thing this does is add the PR numbers to the display? I somehow expected you would be able to do more with this; the obvious wish is to open a PR in the web browser (I have a custom command for this). Maybe the o command could be overloaded for this? Show the existing PR if we know there is one, and open a new one if not.

Yep currently it's just for display but I like your idea.

dlvhdr commented on pkg/commands/git_commands/github.go:
```diff
@@ -0,0 +1,436 @@
+
+func fetchPullRequestsQuery(branches []string, owner string, repo string) string {
+	var queries []string
+	for i, branch := range branches {
```
You can do this in one request (not sure about the length limits for the query) with:

```graphql
query {
  search(first:3, query:"head:gh-integration-3 head:integration-tests-on-windows repo:jesseduffield/lazygit", type: ISSUE) {
    nodes {
      ... on PullRequest {
        url
      }
    }
  }
}
```

Daemoen commented:
Is there a plan to make gh pr usable in LG, or only the external view?  I would *love* to see it allow us to use 'gh pr create/edit/etc'.. if lg could add in support for the gh command for prs... it would be amazing and make it so i dont even have to leave my nvim/lg setup xD

mpasa commented:
@Daemoen You can easily do this as a customCommand. For example, for creating a PR on the current branch:

```
customCommands:
  - key: "o"
    command: "gh pr create"
    subprocess: true
    context: "localBranches"
```

OliverJAsh commented:
Could we show a list of pull requests and their titles? I would actually use this instead of the branch view most of the time. This way I wouldn't need to remember the name of the branch or the PR number. I just need the (human readable) title.

jesseduffield commented:
> branch name ambiguity. In some repos it is common to reuse the same branch name again and again for similar PRs; one example from this repo is the https://github.com/jesseduffield/lazygit/pull/5228 PRs, but there are many others. In such a case, lazygit shows the oldest one of these PR (usually a closed one), which makes no sense at all. It should either use the newest, or find an open one, or some combination of both.

I've got it sorting by newest first now when fetching PRs. Picking the newest should be fine.

> the behavior when using it in a repo for the first time (i.e. when there's no gh-resolved config yet); currently it shows a confusing browser with no title. I wonder if we should just default to 'origin' if there is one, or show no PRs if not; users then have to use gh on the command line to set the base repo.

Now it selects the one repo if there's only one, otherwise 'origin', otherwise prompts the user and I've got a title on the prompt now. I think this works well.

jesseduffield committed:
Refactor hosting service to support repo name extraction

Extract parseRemoteUrl helper to avoid duplication between
getRepoURLFromRemoteURL and new getRepoNameFromRemoteURL.
Add repoNameTemplate to ServiceDefinition and GetRepoName() to
HostingServiceMgr. Add GetRepoInfoFromURL for callers that need
owner/repo without a full service lookup.

jesseduffield committed:
Add cli/go-gh/v2 and cli/safeexec vendor dependencies

Required for authenticating with GitHub's API using the token
stored by the gh CLI.

jesseduffield committed:
Add GitHub commands and model for fetching PR status

Add GitHubCommands struct with GraphQL-based PR fetching, and
GithubPullRequest model. Wire HostingService and GitHub command
structs into GitCommand.

Co-authored-by: Stefan Haller <stefan@haller-berlin.de>

jesseduffield committed:
Show PR icons in branches list

Co-authored-by: Stefan Haller <stefan@haller-berlin.de>

stefanhaller commented on docs-master/Config.md:
```diff
@@ -481,6 +481,10 @@ git:
 
+  # If true and if `gh` is installed and on version >=2, we will use `gh` to
+  # display pull requests against branches.
+  enableGithubCli: true
```
It's funny: claude hallucinated about this in the commit message. But it's right, gui.showPRs feels like the better config name; we should name the config after what it does rather than how it's implemented.

But I wonder if we really need a config at all. What could be a reason to turn it off?

stefanhaller commented on pkg/gui/controllers/helpers/helpers.go:
```diff
@@ -53,6 +52,7 @@ type Helpers struct {
 	Search            *SearchHelper
 	Worktree          *WorktreeHelper
 	SubCommits        *SubCommitsHelper
+	Host              *HostHelper
```
Is there any reason for moving this down to the end? Can we drop this change?

stefanhaller commented on pkg/integration/tests/branch/create_tag.go:
```diff
@@ -20,7 +20,7 @@ var CreateTag = NewIntegrationTest(NewIntegrationTestArgs{
 			Focus().
 			Lines(
-				MatchesRegexp(`\*\s*new-branch`).IsSelected(),
+				MatchesRegexp(`\*.*new-branch`).IsSelected(),
```
These can be dropped as well if we stop showing the icon for branches that don't have PRs.

stefanhaller committed:
Remove branch icons from Branches list

The only information that this carries is whether an entry is a real branch or a
detached head. Detached heads can only be at the top, and they are easy to tell
apart from the other branches by their name, so the icon is not really very
useful.

We are going to show PR icons in this column.

stefanhaller committed:
Show PR information in main view, above the branch log

stefanhaller committed:
Add missing entries to scopeNameMap

stefanhaller committed:
Allow turning off nerd fonts without restarting

Not a terribly important feature for users, probably, but it is useful for
developers when testing a new feature with or without nerd fonts; and it does go
against or policy of all configs being hot-reloadable.

Note that switching from nerd fonts version 2 to 3 is *not* supported without
restarting, but I find this less important.

stefanhaller committed:
Call OnMenuPress(nil) when hitting esc in a menu

OnMenuPress can already deal with the selected item being nil, so this allows us
to add common code to it that is run when cancelling the menu.

stefanhaller committed:
Add an optional onCancel hook for menus

stefanhaller committed:
Prompt only once per session for each repo

If the user hits escape in the "Select base repository for pull requests"
prompt, don't bother them again for this repo at the next refresh.

jesseduffield committed:
Add pull requests to lazygit's model and refresh them

Co-authored-by: Stefan Haller <stefan@haller-berlin.de>

jesseduffield committed:
Cache PRs in AppState so that they appear immediately at startup

Co-authored-by: Stefan Haller <stefan@haller-berlin.de>

stefanhaller committed:
Add commands for opening a Github PR in the browser

For the branches panel we might consider unifying it with the existing `o`
command for creating a PR: it could check if there is a PR already, and open it
if so, or create a new one if not.

However, I also want the command in the local commits panel for the checked out
branch, and there's no existing "Create PR" command there; and the `o` command
opens the selected commit in the browser, so it's unrelated.

stefanhaller committed:
Change "Copy pull request URL to clipboard" command to use existing PR if there is one

stefanhaller committed:
Don't refresh pull requests when checking out a local branch

For esthetic reasons, checking out a branch (or other ref) blocks the UI until
the refresh is done, so it's important that the refresh doesn't do unnecessary
work. Refreshing pull requests is unnecessary (but costly, when waiting for it)
when a branch is checked out that already existed locally. However, it is
required when checking out a remote branch for the first time, so that the PR
icon appears immediately when there is one.

stefanhaller committed:
Add some brief documentation

ManuLinares commented:
Is there currently any way to hide (Closed) PRs?

<img width="121" height="47" alt="image" src="https://github.com/user-attachments/assets/8d39edf1-90bc-4ff7-b204-4b34445e4ca7" />
<img width="223" height="54" alt="image" src="https://github.com/user-attachments/assets/4c31c316-601a-4ba5-b81f-ec8691867444" />

stefanhaller commented:
@ManuLinares No; hiding closed PRs for main branches was discussed above, and I'll probably add this soon (on the assumption that a PR on a main branch is almost always a mistake).

For other branches, it is important to see the PR even if closed or merged, because it tells you that you can now delete the branch.

There is one special case that trips me up every time I run into it: creating a new branch and reusing a name that was used before. Happened to me today, the branch was called `fix-failing-test`, and I used that before, a few years ago. In this case, you push the new branch for the first time and it immediately shows a magenta ("merged") icon next to it, which is quite confusing. But there's no way for lazygit to tell that this is a different branch from the one back then, so I don't see what we can do about it.

ruudk commented:
Doesn't this show that maybe this feature is, flawed? I mean I like it but the requirement that branches have to be unique is just not working.

stefanhaller commented:
@ruudk If you have a suggestion how we can improve it, I'm all ears.

stefanhaller commented:
@ruudk @ManuLinares @shgew Here's a PR for hiding PRs on main branches: #5501. Please check the description to see if the logic makes sense to you all.
